src/runner/mod.rs

```diff
@@ -0,0 +1,224 @@
+use std::{
+    process,
+    sync::{
+        mpsc::{channel, Receiver},
+        Arc,
+    },
+    time::Duration,
+};
+
+use ethers::{
+    providers::{Middleware, Provider},
+    types::{BlockNumber, H256},
+};
+use eyre::Result;
+use tokio::time::sleep;
+
+use crate::{
+    config::{Config, SyncMode},
+    driver::Driver,
+    engine::{Engine, EngineApi, ExecutionPayload, ForkchoiceState, Status},
+};
+
+const TRUSTED_PEER_ENODE: &str = "enode://e85ba0beec172b17f53b373b0ab72238754259aa39f1ae5290e3244e0120882f4cf95acd203661a27c8618b27ca014d4e193266cb3feae43655ed55358eedb06@3.86.143.120:30303?discport=21693";
+
+pub struct Runner {
+    config: Config,
+    sync_mode: SyncMode,
+    checkpoint_hash: Option<String>,
+    shutdown_recv: Arc<Receiver<bool>>,
+}
+
+impl Runner {
+    pub fn from_config(config: Config) -> Self {
+        let (shutdown_sender, shutdown_recv) = channel();
+        ctrlc::set_handler(move || {
+            tracing::info!("shutting down");
+            shutdown_sender
+                .send(true)
+                .expect("could not send shutdown signal");
+        })
+        .expect("could not register shutdown handler");
+
+        Self {
+            config,
+            sync_mode: SyncMode::Full,
+            checkpoint_hash: None,
+            shutdown_recv: Arc::new(shutdown_recv),
+        }
+    }
+
+    pub fn with_sync_mode(mut self, sync_mode: SyncMode) -> Self {
+        self.sync_mode = sync_mode;
+        self
+    }
+
+    pub fn with_checkpoint_hash(mut self, checkpoint_hash: Option<String>) -> Self {
+        self.checkpoint_hash = checkpoint_hash;
+        self
+    }
+
+    pub async fn run(&self) -> Result<()> {
```

Comment:
Can we make this consume self (take `self` instead of `&self`) to ensure that run is only called once?